src/main.rs

```diff
@@ -114,15 +117,35 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
     let server = Server::bind(&address).serve(service);
 
+    let graceful = server.with_graceful_shutdown(shutdown_signal());
+
     info!("Listening on http://{}", address);
 
-    if let Err(why) = server.await {
+    if let Err(why) = graceful.await {
         error!("Fatal server error: {}", why);
     }
 
     Ok(())
 }
 
+async fn shutdown_signal() {
+    #[cfg(windows)]
+    tokio::signal::ctrl_c()
+        .await
+        .expect("failed to install CTRL+C signal handler");
+    #[cfg(unix)]
+    {
+        let mut sigint = signal(SignalKind::interrupt()).expect("failed to install SIGINT handler");
+        let mut sigterm =
+            signal(SignalKind::terminate()).expect("failed to install SIGTERM handler");
+
+        tokio::select! {
+            _ = sigint.recv() => {},
+            _ = sigterm.recv() => {},
+        };
```

Comment:
This feels unfortunate. Can you think of an alternative for this? I can't off the top of my head without pulling in `futures-util`.

Reply:
That's not possible due to the nature of the tokio signal handling module